# Reading the file

In [1]:
import csv
with open ('laptops.csv') as file:
    rows = list(csv.reader(file))
    header = rows[0]
    rows = rows[1:]
    print(header)
    print("\n")
    print(rows[:5])

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']


[['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'], ['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'], ['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'], ['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'], ['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256G

# Inventory Class
Implementing a class to represent the inventory. It get the name of the CSV file as argument and reads it into self.header and self.rows.

In [2]:
class Inventory():
    
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])

In [3]:
inventory = Inventory('laptops.csv')
print(inventory.header)
print("\n")
print(len(inventory.rows))

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']


1303


# Adding a method to the class that retrieves the data for given laptop id

In [4]:
class Inventory():
    # constructor that intializes the header and rows of data
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])
    # a method to retrieve a row that matches a given laptop_id            
    def get_laptop_from_id(self, laptop_id):
        for each in self.rows:
            if laptop_id == each[0]:
                return each        
        return None
            

In [7]:
inventory  = Inventory('laptops.csv')
print(inventory.get_laptop_from_id('3362737'))
print(inventory.get_laptop_from_id('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]
None


# optimising the laptop look up
Python dictionary which facilitates constant time look up and data retrieval is used here.

In [9]:
class Inventory():
    # constructor that intializes the header and rows of data
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])
            #data is stored in dictionary
            self.id_to_row = {}
            for each in self.rows:
                self.id_to_row[each[0]] = each
    # a method to retrieve a row that matches a given laptop_id            
    def get_laptop_from_id(self, laptop_id):
        for each in self.rows:
            if laptop_id == each[0]:
                return each        
        return None
    
    # an improved methid to retrieve laptop id
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None

In [12]:
inventory  = Inventory('laptops.csv')
print(inventory.get_laptop_from_id_fast('3362737'))
print(inventory.get_laptop_from_id_fast('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]
None


# Comparing Performance
Compare the performance of both function for id lookup.

In [15]:
import time                                                      
import random                                                       

ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]

inventory = Inventory('laptops.csv')                               

total_time_no_dict = 0                                              
for identifier in ids:                                             
    start = time.time()                                             
    inventory.get_laptop_from_id(identifier)                      
    end = time.time()                                               
    total_time_no_dict += end - start                              
    
total_time_dict = 0                                               
for identifier in ids:                                              
    start = time.time()                                            
    inventory.get_laptop_from_id_fast(identifier)                   
    end = time.time()                                             
    total_time_dict += end - start                                  
    
print(total_time_no_dict)                                           
print(total_time_dict)
print(total_time_no_dict/total_time_dict)

0.9889552593231201
0.0041654109954833984
237.4208116307023


# Analysis
We got:

0.9842617511749268
0.004242897033691406

We can see a significant improve in performance. If we divide 0.588 by 0.002 we see that the new method is about 237 times faster for this input size.

# Two Laptop Promotion
Write a method that finds whether we can spend a given amount of money by purchasing either one or two laptops.

In [19]:
class Inventory():
    # constructor that intializes the header and rows of data
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])
            #data is stored in dictionary
            self.id_to_row = {}
            for each in self.rows:
                self.id_to_row[each[0]] = each
    # a method to retrieve a row that matches a given laptop_id            
    def get_laptop_from_id(self, laptop_id):
        for each in self.rows:
            if laptop_id == each[0]:
                return each        
        return None
    
    # an improved methid to retrieve laptop id
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    #Given Gift card, check whther it is possible to spend the 
    #enitre amount for atmost two latops
    def check_promotion_dollars(self, dollars):
        for each in self.rows:
            # if gift equals one laptop price
            if dollars == each[-1]:
                return True
            #if two laptop adds uptop one laptop price
        for each1 in self.rows:
            for each2 in self.rows:
                if each1[-1] + each2[-1] == dollars:
                    return True
        return False

In [20]:
inventory = Inventory('laptops.csv')               
print(inventory.check_promotion_dollars(1000))    
print(inventory.check_promotion_dollars(442))

True
False


# Optimizing Laptop Promotion
Create a faster version of the promotion method by using one for loop. Also the data is preprocessed into set() which provides O(1) time complexity during look up but has increased memory complexity correspoding to no of prices data

In [28]:
class Inventory():
    # constructor that intializes the header and rows of data
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])
            #data is stored in dictionary
            self.id_to_row = {}
            for each in self.rows:
                self.id_to_row[each[0]] = each
            #set () facilitates O(1) time complexity for look up
            self.prices = set()
            for each in self.rows:
                self.prices.add(each[-1])
    
    
    # a method to retrieve a row that matches a given laptop_id            
    def get_laptop_from_id(self, laptop_id):
        for each in self.rows:
            if laptop_id == each[0]:
                return each        
        return None
    
    
    # an improved methid to retrieve laptop id
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    #Given Gift card, check whther it is possible to spend the 
    #enitre amount for atmost two latops
    def check_promotion_dollars(self, dollars):
        for each in self.rows:
            # if gift equals one laptop price
            if dollars == each[-1]:
                return True
            #if two laptop adds uptop one laptop price
        for each1 in self.rows:
            for each2 in self.rows:
                if each1[-1] + each2[-1] == dollars:
                    return True
        return False
    
    #a faster method which has one for loop and a data store in set()
    def check_promotion_dollars_fast(self, dollars):
        for each in self.rows:
            # if gift equals one laptop price
            if dollars in self.prices:
                return True
        #if two laptop adds uptop one laptop price and check in set()
        for price  in self.prices:
            if dollars - price in self.prices:
                return True
        return False

        

# Test the code

In [29]:
inventory = Inventory('laptops.csv')                 
print(inventory.check_promotion_dollars_fast(1000))  
print(inventory.check_promotion_dollars_fast(442))

True
False


# Comparing Promotion Functions
Compare the performance of both methods for the promotion.

In [31]:

prices = [random.randint(100, 5000) for _ in range(100)] 

inventory = Inventory('laptops.csv')                     

total_time_no_set = 0                                    
for price in prices:                                     
    start = time.time()                                  
    inventory.check_promotion_dollars(price)             
    end = time.time()                                    
    total_time_no_set += end - start                     
    
total_time_set = 0                                       
for price in prices:                                     
    start = time.time()                                  
    inventory.check_promotion_dollars_fast(price)        
    end = time.time()                                    
    total_time_set += end - start                        
    
print(total_time_no_set)                                 
print(total_time_set)
print(total_time_no_set/total_time_set)

2.6307811737060547
0.008604288101196289
305.75233450635926


# Analysis
We got:

2.6307811737060547
0.008604288101196289
We can see a significant improve in performance. If we divide 2.63078 by 0.0086 we see that the new method is about 305 times faster for this input size.
# Finding Laptops Within a Budget
Implement a method for finding the range of indexes of laptops that fall within a budget.

In [37]:
def row_price(row):
    return row[-1]

class Inventory():
    # constructor that intializes the header and rows of data
    def __init__(self,csv_filename):
        with open(csv_filename, mode = 'r') as file:
            row = list(csv.reader(file))
            self.header = row[0]
            self.rows = row[1:]
            for each in self.rows:
                each[12] = int(each[12])
            #data is stored in dictionary
            self.id_to_row = {}
            for each in self.rows:
                self.id_to_row[each[0]] = each
            #set () facilitates O(1) time complexity for look up
            self.prices = set()
            for each in self.rows:
                self.prices.add(each[-1])
            self.rows_by_price = sorted(self.rows, key=row_price)
            
    
    
    # a method to retrieve a row that matches a given laptop_id            
    def get_laptop_from_id(self, laptop_id):
        for each in self.rows:
            if laptop_id == each[0]:
                return each        
        return None
    
    
    # an improved methid to retrieve laptop id
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    #Given Gift card, check whther it is possible to spend the 
    #enitre amount for atmost two latops
    def check_promotion_dollars(self, dollars):
        for each in self.rows:
            # if gift equals one laptop price
            if dollars == each[-1]:
                return True
            #if two laptop adds uptop one laptop price
        for each1 in self.rows:
            for each2 in self.rows:
                if each1[-1] + each2[-1] == dollars:
                    return True
        return False
    
    #a faster method which has one for loop and a data  in set()
    def check_promotion_dollars_fast(self, dollars):
        for each in self.rows:
            # if gift equals one laptop price
            if dollars in self.prices:
                return True
        #if two laptop adds upto one laptop price and in set()
        for price  in self.prices:
            if dollars - price in self.prices:
                return True
        return False

    #a method to look for laptops below a given price
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0                                   
        range_end = len(self.rows_by_price) - 1                   
        while range_start < range_end:
            range_middle = (range_end + range_start) // 2  
            price = self.rows_by_price[range_middle][-1]
            if price > target_price:
                range_end = range_middle
            else:
                range_start = range_middle + 1
        if self.rows_by_price[range_start][-1] <= target_price:                  
            return -1                                   
        return range_start

In [38]:
inventory = Inventory('laptops.csv')                                
print(inventory.find_first_laptop_more_expensive(1000))  
print(inventory.find_first_laptop_more_expensive(10000)) 

683
-1
